feat(config): add config crate

mattsse opened the pull request:
closes #291
Adds a config crate, see `README` for description

gakonst commented:
I'm +1 on this design, plus other things like compiler optimizations

gakonst commented:
This is already looking really good. One thing we should do is:
1. Touch a default version of it during `forge init`
2. Make `forge` assume that the project root is where the `foundry.toml` is, instead of where the git project root is. It would traverse from the current directory up to the git root directory looking for a foundry.toml, and stop when it finds it.

mattsse commented:
sgtm!

traversing dirs is built in directly to `figment` and is definitely negligible, also considering that cargo does the same thing.
this will probably always be between 4-10 levels and dir lookups are quite fast.

gakonst commented on config/README.md:
```diff
@@ -0,0 +1,76 @@
+If the path set in `FOUNDRY_CONFIG` is absolute, no such search takes place and the absolute path is used directly.
+
+In `foundry.toml` you can define multiple profiles, therefore the file is assumed to be _nested_, so each top-level key
+declares a profile and its values the value for the profile. The following is an example of what such a file might look
```
Some type here in "its values"?

gakonst commented on config/README.md:
```diff
@@ -0,0 +1,76 @@
+eth-rpc-url = "https://mainnet.infura.io"
+
+## set only when the `hardhat` profile is selected
+[hardhat]
```
When somebody uses `--hardhat` in the cli, should this profile be automatically turned on?

mattsse replied:
since this is the cli flag, it should override the config?
but the hardhat style paths are set during init in the `[default]` profile

gakonst commented on config/README.md:
```diff
@@ -0,0 +1,76 @@
+
+The following is a foundry.toml file with all configuration options set.
+
+TODO fill out
```
todo

gakonst commented on config/src/lib.rs:
```diff
@@ -0,0 +1,482 @@
+    pub libs: Vec<PathBuf>,
+    /// whether to enable cache
+    pub cache: bool,
+    /// concrete solc version to use if any,
+    pub solc_version: Option<Version>,
```
So this would trigger --no-autodetect?

mattsse replied:
yeah but it's setting solc version and `autodetect` some what overlap,
so I think solc_version should take precedence over `auto-detect`.
by default autodetect is true.

gakonst commented on config/src/lib.rs:
```diff
@@ -0,0 +1,482 @@
+    /// library addresses to link
+    pub libraries: Vec<Address>,
+
+    /// PRIVATE: This structure may grow, As such, constructing this structure should
+    /// _always_ be done using a public constructor or update syntax:
```
Hm typically you'd do this with all fields private, and only exposing getters, maybe worth using something like https://crates.io/crates/derive-getters?

mattsse replied:
this enforces that we always use this pattern with the trailing `..Config::default()`, so that, if we add new config vars we don't end up with `missing field` values

```rust
Config {
    src: "contracts".into(),
    out: "artifacts".into(),
    libs: vec!["node_modules".into()],
    ..Config::default()
}
```

gakonst commented on config/src/lib.rs:
````diff
@@ -0,0 +1,482 @@
+    /// ```
+    pub fn with_root(root: impl Into<PathBuf>) -> Self {
+        // autodetect paths
+        let paths = ProjectPathsConfig::builder().build_with_root(root);
````
👍

gakonst commented on config/src/lib.rs:
```diff
@@ -0,0 +1,482 @@
+    }
+
+    /// Returns the default config but with hardhat paths
+    pub fn hardhat() -> Self {
```
Should we have a `fn dapptools()` too which aliases to `fn default()`?

mattsse replied:
that's what `Config::default` uses

gakonst reviewed (approved):
LGTM so far, minor nits. Just need to:
1. add it in `forge init`
2. add some docs about it in the top-level README, to guide ppl to start creating these config files in their dev envs
3. change the root dir detection from being the .git dir, to the first directory `foundry.toml` is found in OR the .git dir for a project.

mattsse commented:
# Updates
* extended config values and added a complete example in `config/README`
* support and optimize relative `Remappings` in foundry.toml, so that remappings can be added as `remappings = ["ds-test=lib/ds-test/"]`, starting from the project root.
* updated main readme
* added `forge config` command to show the currently used values
* include `BasicConfig` in `forge init`
* add integration test for init

@gakonst

mattsse commented:
Todo
- [ ] integrate `Config::load` in all commands

gakonst commented on cli/src/cmd/init.rs:
```diff
@@ -0,0 +1,88 @@
+            Dependency::from_str("https://github.com/dapphub/ds-test")
+                .and_then(|dependency| install(&root, vec![dependency]))?;
+
+            // write foundry.toml
+            let config = Config::from(Config::with_root(&root)).into_basic();
+            let dest = root.join(Config::FILE_NAME);
+            if !dest.exists() {
+                std::fs::write(dest, config.to_string_pretty()?)?;
+            }
```
woohooo. Should the `config` instantiation be done inside the `if !dest` branch?

gakonst commented on config/src/lib.rs:
```diff
@@ -0,0 +1,808 @@
+    }
+}
+
+/// The path part of the [`Remapping`] that knows the path of the file it was configured in, if any.
+///
+/// A [`Remapping`] is intended to be absolute, but paths in configuration files are often desired
+/// to be relative to the configuration file itself. For example, a path of
+/// `weird-erc20/=lib/weird-erc20/src/` configured in a file `/var/foundry.toml` might be desired to
+/// resolve as a `weird-erc20/=/var/lib/weird-erc20/src/` remapping.
+#[derive(Debug, Clone, PartialEq)]
+pub struct RelativeRemappingPathBuf {
+    parent: Option<PathBuf>,
+    path: PathBuf,
+}
```
Should this logic be merged with the `Remapping` struct?

mattsse replied:
+1 on moving this to ethers-solc

onbjerg commented:
Not sure if it's out of scope for this PR, but should we also have something like `.sethrc` (`cast.toml`?) to configure things like `ETH_FROM` etc. without having to explicitly state it every time? [`.sethrc` example](https://github.com/dapphub/dapptools/blob/master/src/seth/README.md#example-sethrc-file)

gakonst commented:
I'd pull all of these Configs in the Foundry.toml, instead of having many configs

mattsse commented:
yeah we can support a `[cast]` section and a `[foundry]` where you'd set some meta config vals like color palette

mattsse commented on config/README.md:
````diff
@@ -0,0 +1,104 @@
+```toml
+## defaults for _all_ profiles
+[default]
+src = 'src'
+test = 'test'
````
both would result in the same if we check with `Path::has_root`

```rust
let src_dir = if config.src.has_root() {
     config.src
} else {
   root.join(config.src)
}

```

mattsse commented on config/README.md:
```diff
@@ -0,0 +1,104 @@
+test = 'test'
+out = 'out'
+libs = ['lib']
+remappings = []
+libraries = []
```
oh right, they would, so I think they should be optional in the basic config and only be set if we found some with `forge init`

mattsse commented on config/README.md:
```diff
@@ -0,0 +1,104 @@
+optimizer = false
+optimizer_runs = 200
+verbosity = 0
+ignored_error_codes = []
```
yeah, seems like a good thing to have in ethers-solc, perhaps there is way to figure out what error codes are propagated to solc, will take another look at solidity codebase

gakonst commented on cli/src/cmd/build.rs:
```diff
@@ -251,6 +260,25 @@ impl BuildArgs {
+        Metadata::named("Build Args Provider")
+    }
+
+    fn data(&self) -> Result<Map<Profile, Dict>, Error> {
+        let mut dict = Dict::new();
+
+        if let Some(ref src) = self.contracts {
+            dict.insert("src".to_string(), Value::from(format!("{}", src.display())));
+        }
+
+        // TODO remaining configs
+
+        Ok(Map::from([(Config::selected_profile(), dict)]))
+    }
```
Is the TODO for this PR, or for a follow up?

FWIW me doing `forge install` does not populate the config file with the new remapping, and `forge remappings` doesn't read any remappings from the config file. I guess that's intended right?

mattsse replied:
still todo for this, is close now.

forge install probably should add them by default, that should be possible.

hmm, perhaps we let `forge remappings` always autodetect everything by scanning the libs, should be more useful then reading the config I guess

gakonst commented on cli/src/cmd/build.rs:
```diff
@@ -15,10 +15,19 @@ use std::{
 };
 
 use crate::{cmd::Cmd, opts::forge::CompilerArgs, utils};
+use foundry_config::{
+    figment::{Error, Metadata, Profile, Provider},
+    Config,
+};
```
Should prob combine the 2 `foundry_config::figment` imports?

gakonst reviewed (commented):
Seems like this is also a good opportunity to solve the `--root .` issue in monorepos? Would need to add the path traversal logic to find the foundry.toml

mattsse commented:
> Rebased / fixed conflicts.

🙏

> Seems like this is also a good opportunity to solve the `--root .` issue in monorepos? Would need to add the path traversal logic to find the foundry.toml

yeah, I think it should be

look for a `foundry.toml` file starting at cwd, but only until we see a .git repository so that

```
-- foundry.toml

-- repo
  |__ .git
  |__sub
     |__ cwd

```

will still detect `repo` as root

mattsse commented:
requires last refactorings and some integration tests

mattsse commented:
# Update 
* Simplified `BuildArgs`
* Added separate CLI arg types `EvmArgs` instead of using `EvmOpts`
  *  All values in `EvmArgs` are optional so that they only override the if they are present as CLI argument. But all have a default in `Config` so we can always safely extract an `EvmOpts`
* integrated `Figment`/`Config` in the subcommand by implementing `figment::Provider` on the args
* More docs on how `Figment` and `Config` works
* more remapping support:
  * CLI > ENV > remappings.txt > autodetect
  * deviated from previous approach that merged all (Env + remappings.txt)

# Todo
- [ ] additional tests to ensure values are set correctly 

cc @gakonst
